### DISCUSSION TITLE
Remove white parts of the image while keeping low saturation colors intact

### DISCUSSION THREAD

**Discussion by baykalokandemir** [ImageMagick/ImageMagick#7810] [Help] [answered]
### ImageMagick version

7.1.1-40

### Operating system, version and so on

Windows 11

### Description

Hi! I'm trying to come up with a pipeline to edit screenshots so that everything looks good on my editor in dark mode. This is the pipeline I have so far:

```
REM Step 1: Remove the white background and make it transparent
magick "%input%" -fuzz 15%% -transparent white "%dir%%filename%_step1.png"

REM Step 2: Enhance the saturation and brightness of the colored shape
magick "%dir%%filename%_step1.png" -modulate 110,120,100 "%dir%%filename%_step2.png"

REM Step 3: Upscale the image for better processing
magick "%dir%%filename%_step2.png" -resize 400%% "%dir%%filename%_step3.png"

REM Step 4: Replace black text with white
magick "%dir%%filename%_step3.png" -fuzz 30%% -fill white -opaque black "%dir%%filename%_step4.png"

REM Step 5: Downscale the image back to the original size
magick "%dir%%filename%_step4.png" -resize 25%% "%output%"

REM Step 6: Make colors pastel-like (lighter and vibrant)
magick "%output%" -modulate 100,150,100 -level 90%,100% -gamma 2.2 -colorspace RGB "%output%"
```

My problem arises at step 1. 15% fuzz is just not good enough to cleanly delete all the white surrounding text/shapes on lower resolution images. Unfortunately, I can't really get these images in higher resolutions, so I'm stuck having to choose between having good background transparency vs. having light colors. This illustrates the issue of colors disappearing somewhat well:

Original:
![Pasted image 20241119184552](https://github.com/user-attachments/assets/6bfbf1f9-7682-4b2a-aba6-77717f4d00f6)

15% fuzz:
![Pasted image 20241119184552_processed](https://github.com/user-attachments/assets/954ef2c4-4cf2-46dc-a19f-41b1cc1302c9)

40% fuzz:
![Pasted image 20241119184552_processed](https://github.com/user-attachments/assets/e819e9a3-d8fc-40f3-8902-a403766805ae)

And this illustrates the need for higher fuzz pretty well:

Original:
![Pasted image 20241115145946](https://github.com/user-attachments/assets/43a62e42-4331-4c99-bbf4-65381d02edab)

15% fuzz:
![Pasted image 20241115145946_processed](https://github.com/user-attachments/assets/cfef543e-c893-41f0-a290-84f52521c0aa)

40% fuzz:
![Pasted image 20241115145946_processed](https://github.com/user-attachments/assets/d284fd45-8a0a-474b-b51e-f21aabd54de0)

Any ideas on how to improve / replace this background removal procedure would be greatly appreciated!

**Comment by fmw42** [comment]
Try this (threshold to create mask for the white area, negate it and put that image into the alpha channel of the input)

`magick image.png -alpha off -write mpr:image -negate -threshold 5%% mpr:image +swap -alpha off -compose copy_opacity -composite result.png`

![result](https://github.com/user-attachments/assets/5370f988-be9a-4a4c-aba1-97a9b859f20a)


Adjust the threshold to remove more or less white around the black letters and lines. For windows I have doubled the % to %%.

**Comment by snibgo** [comment, marked as answer]
Where inpcodeout.png is your supplied image, I suggest this Windows BAT command:
```
magick ^
  inpcodeout.png ^
  -colorspace HCL ^
  -channel 2 ^
    -negate ^
    -evaluate Pow 0.7 ^
  -channel 1 ^
    -evaluate Multiply 1.2 ^
  +channel ^
  -write mpr:hicl ^
  -colorspace sRGB ^
  -write x1.png ^
  -write mpr:Coloured ^
  +delete ^
  mpr:hicl ^
  -channel 1 -threshold 0 ^
  -channel 1,2 ^
  -separate ^
  -write x2.png ^
  -compose Lighten -composite ^
  -write m.png ^
  -write mpr:Alpha ^
  -delete 0--1 ^
  mpr:Coloured ^
  mpr:Alpha ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  out.png
```

Each "-write x?.png" writes an image we can use for debugging. When this does what you want, you can remove all these lines.

This works in HCL colorspace. We negate channel 2 which is lightness, and "-evaluate Pow 0.7" makes it lighter (use a lower number to make it lighter still, as you like). We multiply channel 1 (chroma) to increase saturation, as much as you like. Convert to sRGB and save in mpr:Coloured. This will be the new colour channels.

From the HCL image, channel 1 (chroma) threshold at 0 so all pixels with any chroma are set to maximum. Separate this and the Lightness, and take the maximum of these two ("-compose Lighten -composite"). This will be the new alpha channel.

CopyOpacity the new alpha to the new colour channels. The result out.png is:

![out](https://github.com/user-attachments/assets/ce8f896a-d2c4-4e80-b590-c45e40a29f1d)

**Reply by baykalokandemir** [reply]
Hi! Sorry I took so long to answer, this is an amazing solution! It works great for my purposes. There's some fringing on lower resolution / very thin text, but it's infinitely more readable. Thanks for your help!

**Comment by snibgo** [comment]
Yes, my command above leaves gray anti-aliased partially transparent pixels around the white lines and text. So when the image is flattened on a white (or nearly white) background, the white lines and text are still visible, because they have gray borders. So I think this is useful. (A feature, not a bug!)

However, if you don't like this, we can change all pixels that are partially or entirely transparent to be white, without changing the transparency. Add three lines at the end of the command:
```
magick ^
  inpcodeout.png ^
  -colorspace HCL ^
  -channel 2 ^
    -negate ^
    -evaluate Pow 0.7 ^
  -channel 1 ^
    -evaluate Multiply 1.2 ^
  +channel ^
  -write mpr:hicl ^
  -colorspace sRGB ^
  -write x1.png ^
  -write mpr:Coloured ^
  +delete ^
  mpr:hicl ^
  -channel 1 -threshold 0 ^
  -channel 1,2 ^
  -separate ^
  -write x2.png ^
  -compose Lighten -composite ^
  -write m.png ^
  -write mpr:Alpha ^
  -delete 0--1 ^
  mpr:Coloured ^
  mpr:Alpha ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  -channel RGB ^
  -fx "a<1?1:u" ^
  +channel ^
  out2.png
```
